programs/drift/src/instructions/lp_pool.rs

```diff
@@ -0,0 +1,1554 @@
+        perp_market::{AmmCacheFixed, CacheInfo, AMM_POSITIONS_CACHE},
+        perp_market_map::MarketSet,
+        spot_market::{SpotBalanceType, SpotMarket},
+        spot_market_map::get_writable_spot_market_set_from_many,
+        state::State,
+        user::MarketType,
+        zero_copy::{AccountZeroCopy, AccountZeroCopyMut, ZeroCopyLoader},
+    },
+    validate,
+};
+
+use solana_program::sysvar::clock::Clock;
+
+use super::optional_accounts::{load_maps, AccountMaps};
+use crate::controller::spot_balance::update_spot_market_cumulative_interest;
+use crate::controller::token::{receive, send_from_program_vault};
+use crate::instructions::constraints::*;
+use crate::state::lp_pool::{
+    AMM_MAP_PDA_SEED, CONSTITUENT_PDA_SEED, CONSTITUENT_TARGET_BASE_PDA_SEED,
+    LP_POOL_TOKEN_VAULT_PDA_SEED,
+};
+
+pub fn handle_update_constituent_target_base<'c: 'info, 'info>(
+    ctx: Context<'_, '_, 'c, 'info, UpdateConstituentTargetBase<'info>>,
+) -> Result<()> {
+    let slot = Clock::get()?.slot;
+
+    let lp_pool = &ctx.accounts.lp_pool.load()?;
+    let lp_pool_key: &Pubkey = &ctx.accounts.lp_pool.key();
+    let amm_cache_key: &Pubkey = &ctx.accounts.amm_cache.key();
+
+    let amm_cache: AccountZeroCopy<'_, CacheInfo, AmmCacheFixed> =
+        ctx.accounts.amm_cache.load_zc()?;
+
+    amm_cache.check_oracle_staleness(slot, MAX_AMM_CACHE_STALENESS_FOR_TARGET_CALC)?;
+    amm_cache.check_perp_market_staleness(slot, MAX_AMM_CACHE_STALENESS_FOR_TARGET_CALC)?;
+
+    let expected_cache_pda = &Pubkey::create_program_address(
+        &[
+            AMM_POSITIONS_CACHE.as_ref(),
+            amm_cache.fixed.bump.to_le_bytes().as_ref(),
+        ],
+        &crate::ID,
+    )
+    .map_err(|_| ErrorCode::InvalidPDA)?;
+    validate!(
+        expected_cache_pda.eq(amm_cache_key),
+        ErrorCode::InvalidPDA,
+        "Amm cache PDA does not match expected PDA"
+    )?;
+
+    let state = &ctx.accounts.state;
+    let constituent_target_base_key = &ctx.accounts.constituent_target_base.key();
+    let amm_mapping_key = &ctx.accounts.amm_constituent_mapping.key();
+
+    let mut constituent_target_base: AccountZeroCopyMut<
+        '_,
+        TargetsDatum,
+        ConstituentTargetBaseFixed,
+    > = ctx.accounts.constituent_target_base.load_zc_mut()?;
+
+    let bump = constituent_target_base.fixed.bump;
+    let expected_pda = &Pubkey::create_program_address(
+        &[
+            CONSTITUENT_TARGET_BASE_PDA_SEED.as_ref(),
+            lp_pool.pubkey.as_ref(),
+            bump.to_le_bytes().as_ref(),
+        ],
+        &crate::ID,
+    )
+    .map_err(|_| ErrorCode::InvalidPDA)?;
+    validate!(
+        expected_pda.eq(constituent_target_base_key),
+        ErrorCode::InvalidPDA,
+        "Constituent target weights PDA does not match expected PDA"
+    )?;
+
+    let num_constituents = constituent_target_base.len();
+    for datum in constituent_target_base.iter() {
+        msg!("weight datum: {:?}", datum);
```

Comment:
need to do an audit of all the msg to figure out which ones we want vs arent useful